airbyte-scheduler/src/main/java/io/airbyte/scheduler/worker_run/WorkerRunWithEnvironmentFactory.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2020 Airbyte
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.scheduler.worker_run;
+
+import io.airbyte.workers.process.ProcessBuilderFactory;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The constructor takes in information from the environment (e.g. Scheduler versus Temporal) in
+ * which the job will run. This information (e.g. file system roots and pbfs) can be used to create
+ * any job of a certain type. It also takes in a factory for creating a WorkerRun for a specific job
+ * type. The create method takes in the configuration information for a specific instance of a job
+ * run.
+ *
+ * @param <T> type of the input config.
+ */
+public class WorkerRunWithEnvironmentFactory<T> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(WorkerRunWithEnvironmentFactory.class);
+
+  private final Path workspaceRoot;
+  private final ProcessBuilderFactory pbf;
+  private final WorkerRunFactory<T> workerRunFactory;
+
+  public WorkerRunWithEnvironmentFactory(final Path workspaceRoot,
+                                         final ProcessBuilderFactory pbf,
+                                         final WorkerRunFactory<T> workerRunFactory) {
+    this.workspaceRoot = workspaceRoot;
+    this.pbf = pbf;
+    this.workerRunFactory = workerRunFactory;
+  }
+
+  public WorkerRun create(final long jobId, final int attempt, final T config) {
```

Comment:
I assume this logic of creating path can at some point belong to the workerrun right (once we are 100% on temporal)? it is a bit overkill at the moment to have it.

Reply:
definitely.

Reply:
i think we can redefine the worker run concept a bit going forward to simplify this, but i think figuring that out will be easier once we're on temporal because we'll know exactly what the constrains are.